File: src/main/java/com/citylogic/domain/map/Cell.java
package com.citylogic.domain.map;

import com.citylogic.domain.buildings.BuildingInstance;
import com.citylogic.domain.core.Point;

/**
 * Individual tile/cell in the city matrix.
 */
public class Cell {
    private final int x;
    private final int y;
    private final Point position;
    private double pollutionLevel;
    private BuildingInstance currentBuilding;

    public Cell(int x, int y) {
        this.x = x;
        this.y = y;
        this.position = new Point(x, y);
        this.pollutionLevel = 0.0;
        this.currentBuilding = null;
    }

    public int getX() {
        return x;
    }

    public int getY() {
        return y;
    }

    public Point getPosition() {
        return position;
    }

    public boolean isOccupied() {
        return currentBuilding != null;
    }

    public BuildingInstance getBuilding() {
        return currentBuilding;
    }

    public void setBuilding(BuildingInstance building) {
        this.currentBuilding = building;
    }

    public void clear() {
        this.currentBuilding = null;
    }

    public double getPollutionLevel() {
        return pollutionLevel;
    }

    public void setPollutionLevel(double pollutionLevel) {
        this.pollutionLevel = Math.max(0.0, pollutionLevel);
    }
}

File: src/domain/map/Cell.ts
import { Point } from '../../types';
import { BuildingInstance } from '../buildings/BuildingInstance';

export class Cell {
  public readonly x: number;
  public readonly y: number;
  public readonly position: Point;
  public pollutionLevel: number;
  public currentBuilding: BuildingInstance | null;

  constructor(x: number, y: number) {
    this.x = x;
    this.y = y;
    this.position = { x, y };
    this.pollutionLevel = 0;
    this.currentBuilding = null;
  }

  public isOccupied(): boolean {
    return this.currentBuilding !== null;
  }

  public setBuilding(building: BuildingInstance): void {
    this.currentBuilding = building;
  }

  public clear(): void {
    this.currentBuilding = null;
  }

  public getBuilding(): BuildingInstance | null {
    return this.currentBuilding;
  }

  public getPollutionLevel(): number {
    return this.pollutionLevel;
  }
}
